docs/source/format/Metadata.rst

```diff
@@ -390,7 +387,36 @@ Time supports the same unit resolutions: second, millisecond, microsecond, and
 nanosecond. We represent time as the smallest integer accommodating the
 indicated unit. For second and millisecond: 32-bit, for the others 64-bit.
 
-Dictionary encoding
--------------------
+Custom Application Metadata
+---------------------------
+
+We provide a ``custom_metadata`` field at three levels to provide a
+mechanism for developers to pass application-specific metadata in
+Arrow protocol messages. This includes ``Field``, ``Schema``, and
+``Message``.
+
+
+The ``ARROW:`` pattern (colon included) is a reserved prefix for
+internal Arrow use in the ``custom_metadata`` fields.
+
+Extension Types
+---------------
+
+User-defined "extension" types can be defined setting certain
+``KeyValue`` pairs in ``custom_metadata` in the ``Field`` metadata
+structure. These extension keys are:
+
+* ``'ARROW:extension_name'`` for the string name identifying the
+  custom data type
+* ``'ARROW:extension_data'`` for a serialized representation of the
```

Comment:
should this be called metadata, to be clear it should be small?  I asked this on the ML but could you provide an example for how this would be used for Tensors?  (and clarify UUIDs wouldn't use it?)